### PR TITLE
22204 add validation on georeference verification status

### PR DESCRIPTION
**Pull request by luusteve**
https://redmine.biodiversity.agr.gc.ca/issues/22204

### CODE REVIEW

File: src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java [hunk 1]
```diff
@@ -7,31 +7,37 @@
 import org.springframework.validation.Validator;
 
 import ca.gc.aafc.collection.api.entities.GeoreferenceAssertion;
+import ca.gc.aafc.collection.api.entities.GeoreferenceAssertion.GeoreferenceVerificationStatus;
 
 @Component
 public class GeoreferenceAssertionValidator implements Validator {
 
-    private MessageSource messageSource;
+  private MessageSource messageSource;
```
Comment (Jkeyuk) [inline src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java:15]: can be final

File: src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java [hunk 2]
```diff
@@ -7,31 +7,37 @@
...
-    }
+  public GeoreferenceAssertionValidator(MessageSource messageSource) {
+    this.messageSource = messageSource;
+  }
 
-    @Override
-    public void validate(Object target, Errors errors) {
-        GeoreferenceAssertion georeferenceAssertion = (GeoreferenceAssertion) target;
+  @Override
+  public boolean supports(Class<?> clazz) {
```
Comment (Jkeyuk) [inline src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java:22]: Overides and should carry the `@NonNull`. like `public boolean supports(@NonNull Class<?> clazz)`

File: src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java [hunk 3]
```diff
@@ -7,31 +7,37 @@
...
+    return GeoreferenceAssertion.class.isAssignableFrom(clazz);
+  }
 
-        if (georeferenceAssertion.getCollectingEvent() != null && georeferenceAssertion.getCollectingEvent().getPrimaryGeoreferenceAssertion() == null) {
-            String errorMessage = messageSource.getMessage("georeferenceAssertion.collectingEvent.primaryGeoreferenceAssertion.null",
-                null, LocaleContextHolder.getLocale());
-            errors.rejectValue("collectingEvent", "georeferenceAssertion.collectingEvent.primaryGeoreferenceAssertion.null", errorMessage);
+  @Override
+  public void validate(Object target, Errors errors) {
```
Comment (Jkeyuk) [inline src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java:27]: Overides and should carry the `@NonNull` ` public void validate(@NonNull Object target, @NonNull Errors errors)`

File: src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java [hunk 4]
```diff
@@ -7,31 +7,37 @@
...
+    GeoreferenceAssertion georeferenceAssertion = (GeoreferenceAssertion) target;
```
Comment (Jkeyuk) [inline src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java:28]: We should probably check that the target is class is supported before casting. that support method seems useful in this case.

File: src/test/java/ca/gc/aafc/collection/api/entities/GeoreferenceAssertionCRUDIT.java [hunk 5]
```diff
@@ -47,7 +50,6 @@ public void testFind() {
         .dwcDecimalLongitude(45.01)
         .dwcCoordinateUncertaintyInMeters(10)
         .dwcGeoreferencedDate(testGeoreferencedDate)
-        .dwcGeoreferenceVerificationStatus(georeferenceVerificationStatus)
         .build();
```
Comment (Jkeyuk) [inline src/test/java/ca/gc/aafc/collection/api/entities/GeoreferenceAssertionCRUDIT.java:53]: not sure why we are removing this from the test case, this should be a valid test of the entity

Comment (Jkeyuk) [reply src/test/java/ca/gc/aafc/collection/api/entities/GeoreferenceAssertionCRUDIT.java:53]: additionally georeferenceVerificationStatus is never used after these changes

**Review by Jkeyuk — CHANGES_REQUESTED** [review]

File: src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java (hunk 4 above)
Comment (luusteve) [reply src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java:28]: that is what the isASsignableFrom in the supports method currently does 
https://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#isAssignableFrom%28java.lang.Class%29

Comment (luusteve) [reply src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java:28]: the supports method is run before the validate method

File: src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java (hunk 2 above)
Comment (luusteve) [reply src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java:22]: Should I modify the CollectingEventValidator to also carry the @NonNull in it's params?

**Comment by luusteve** [comment]
For some reason, the PR is 1 commit behind the action branch. Maybe give it a while before merging

File: src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java (hunk 2 above)
Comment (Jkeyuk) [reply src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java]: I am not sure at this time.

File: src/test/java/ca/gc/aafc/collection/api/entities/GeoreferenceAssertionCRUDIT.java [hunk 6]
```diff
@@ -58,7 +61,6 @@ public void testFind() {
     assertEquals(45.01, fetchedGeoreferenceAssertion.getDwcDecimalLongitude());
     assertEquals(10, fetchedGeoreferenceAssertion.getDwcCoordinateUncertaintyInMeters());
     assertEquals(testGeoreferencedDate, fetchedGeoreferenceAssertion.getDwcGeoreferencedDate());
-    assertEquals(georeferenceVerificationStatus, fetchedGeoreferenceAssertion.getDwcGeoreferenceVerificationStatus());
```
Comment (Jkeyuk) [inline src/test/java/ca/gc/aafc/collection/api/entities/GeoreferenceAssertionCRUDIT.java:61]: I think we should add this back.

File: src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java (hunk 4 above)
Comment (Jkeyuk) [reply src/main/java/ca/gc/aafc/collection/api/validation/GeoreferenceAssertionValidator.java:29]: This appears to be called directly from the CollectingEventService on line 62, additionaly I would be a bit worried as anybody can inject this into a class and call it directly.

Probably safe to just check before we cast.

**Review by Jkeyuk — CHANGES_REQUESTED** [review]

**Pull request merged**